src/taskorchestrationexecutor.ts

```diff
@@ -0,0 +1,493 @@
+        this.exception = undefined;
+        this.orchestratorReturned = false;
+    }
+
+    /**
+     * @hidden
+     *
+     * Start an orchestration's execution, replaying based on the currently-available History.
+     *
+     * @param context
+     *  The orchestration context
+     * @param history
+     *  The orchestration history
+     * @param schemaVersion
+     *  The OOProc output schema version expected by the DF extension
+     * @param fn
+     *  The user-code defining the orchestration
+     *
+     * @returns
+     *  Returns void but communicates the resulting orchestrator state via the context object's handler
+     */
+    public async execute(
+        context: IOrchestrationFunctionContext,
+        history: HistoryEvent[],
+        schemaVersion: UpperSchemaVersion,
+        fn: (context: IOrchestrationFunctionContext) => IterableIterator<unknown>
+    ): Promise<void> {
+        this.schemaVersion = schemaVersion;
+        this.context = context.df;
+        this.generator = fn(context) as Generator<TaskBase, any, any>; // TODO: what happens if code is not a generator?
+
+        // Execute the orchestration, using the history for replay
+        for (const historyEvent of history) {
+            this.processEvent(historyEvent);
+            if (this.hasExecutionCompleted()) {
+                break;
+            }
+        }
+
+        // Construct current orchestration state
+        const actions: IAction[][] = this.actions.length == 0 ? [] : [this.actions];
+        const orchestratorState = new OrchestratorState({
+            isDone: this.orchestrationInvocationCompleted(),
+            actions: actions,
+            output: this.output,
+            error: this.exception?.message,
+            customStatus: this.context.customStatus,
+            schemaVersion: this.schemaVersion,
+        });
+
+        // Record errors, if any
+        let error = undefined;
+        let result: any = orchestratorState;
+        if (this.exception !== undefined) {
+            error = new OrchestrationFailureError(this.orchestratorReturned, orchestratorState);
+            result = undefined;
+        }
+
+        // Communicate the orchestration's current state
+        context.done(error, result);
+        return;
+    }
+
+    /**
+     * @hidden
+     * Determine if the orchestrator should exit, either successfully or through an error.
+     *
+     * @returns
+     *  True if the orchestration's invocation completed, or if an unhandled exception was thrown.
+     *  False otherwise.
+     */
+    private hasExecutionCompleted(): boolean {
+        return this.orchestrationInvocationCompleted() || this.exception !== undefined;
+    }
+
+    /**
+     * @hidden
+     * Determine if the current invocation has finished.
+     *
+     * @returns
+     *  True if the orchestration reached a `return` statement, or a `continueAsNew`.
+     *  False otherwise.
+     */
+    private orchestrationInvocationCompleted(): boolean {
```

Comment:
nit: I suggest coming up with more precise names for `hasExecutionCompleted` and `orchestrationInvocationCompleted`. An "invocation" and an "execution" don't have clearly distinguished meanings. Perhaps "isDoneExecuting()" and "completedSuccessfully()"?